integrations/spark/src/main/java/marquez/spark/agent/lifecycle/plan/LogicalRelationVisitor.java

```diff
@@ -0,0 +1,142 @@
+ * connection URL (schema and authority only) and the table name to be the <code>&lt;database&gt;
+ * </code>.<code>&lt;tableName&gt;</code>.
+ *
+ * <p>{@link org.apache.spark.sql.catalyst.catalog.CatalogTable}s, if present, can be used to
+ * describe the {@link Dataset} if its {@link org.apache.spark.sql.sources.BaseRelation} is unknown.
+ *
+ * <p>If the {@link org.apache.spark.sql.sources.BaseRelation} is unknown, we send back a {@link
+ * Dataset} named for the node name in the logical plan. This helps track what nodes are yet
+ * unknown, while hopefully avoiding gaps in the lineage coverage by providing what information we
+ * have about the dataset.
+ *
+ * <p>TODO If a user specifies the {@link JDBCOptions#JDBC_QUERY_STRING()} option, we do not parse
+ * the sql to determine the specific tables used. Since we return a List of {@link Dataset}s, we can
+ * parse the sql and determine each table referenced to return a complete list of datasets
+ * referenced.
+ */
+@Slf4j
+public class LogicalRelationVisitor extends AbstractPartialFunction<LogicalPlan, List<Dataset>> {
+  private final SparkContext context;
+  private final String jobNamespace;
+
+  public LogicalRelationVisitor(SparkContext context, String jobNamespace) {
+    this.context = context;
+    this.jobNamespace = jobNamespace;
+  }
+
+  @Override
+  public boolean isDefinedAt(LogicalPlan x) {
+    return x instanceof LogicalRelation
+        &&
+        // ignore DatasetSources since they're handled by the DatasetSourceVisitor
+        !(((LogicalRelation) x).relation() instanceof DatasetSource);
+  }
+
+  @Override
+  public List<Dataset> apply(LogicalPlan x) {
+    LogicalRelation logRel = (LogicalRelation) x;
+    if (logRel.relation() instanceof HadoopFsRelation) {
+      return handleHadoopFsRelation((LogicalRelation) x);
+    } else if (logRel.relation() instanceof JDBCRelation) {
+      return handleJdbcRelation((LogicalRelation) x);
+    } else if (logRel.catalogTable().isDefined()) {
+      return handleCatalogTable(logRel);
+    } else {
+      // make a best attempt at capturing the dataset information
+      log.warn("Don't know how to extract dataset from unknown relation {}", logRel.relation());
+      return Collections.singletonList(
+          Dataset.builder()
+              .namespace(jobNamespace)
+              .name(
+                  logRel.relation().getClass().getSimpleName()
+                      + "_"
+                      + logRel.relation().schema().catalogString())
+              .facets(
+                  DatasetFacet.builder()
+                      .description(logRel.simpleString())
+                      .schema(PlanUtils.schemaFacet(logRel.schema()))
+                      .build())
+              .build());
+    }
+  }
+
+  private List<Dataset> handleCatalogTable(LogicalRelation logRel) {
+    CatalogTable catalogTable = logRel.catalogTable().get();
```

Comment:
It is _safe_ to assume the catalog table would be present?

Reply:
We check that it is defined on line 73